adapters/repos/db/vector/cache/paginated_cache.go

```diff
@@ -0,0 +1,505 @@
+	vc.watchForPrune()
+	return vc
+}
+
+func (s *paginatedCache[T]) All() [][]T {
+	s.shardedLocks.RLockAll()
+	defer s.shardedLocks.RUnlockAll()
+
+	out := make([][]T, 0, s.Len())
+	for _, page := range s.cache {
+		for _, vec := range page {
+			if vec != nil {
+				out = append(out, vec)
+			}
+		}
+	}
+
+	return out
+}
+
+// getPageForID returns the page and index of the vector for the given id.
+// This function assumes that the caller holds a read lock on the page.
+func (s *paginatedCache[T]) getPageForID(id uint64) (page [][]T, idx int) {
+	pageIdx := id / PageSize
+	page = s.cache[pageIdx]
+	if page == nil {
+		return nil, -1
+	}
+
+	return page, int(id % PageSize)
+}
+
+// upsert gets or creates a page and applies an update to it.
+// No locks must be held when calling this function.
+func (s *paginatedCache[T]) upsert(id uint64, update func(page [][]T, idx int) error) error {
+	// fast path
+	s.shardedLocks.Lock(id)
+	page, idx := s.getPageForID(id)
+	if page != nil {
+		err := update(page, idx)
+		s.shardedLocks.Unlock(id)
+		return err
+	}
+	s.shardedLocks.Unlock(id)
+
+	s.pageLock.Lock()
+	defer s.pageLock.Unlock()
+
+	// try again in case the page was created in the meantime.
+	// this is to avoid the waiting routines to all acquire a LockAll
+	// lock when multiple goroutines try to create the same page.
+	s.shardedLocks.Lock(id)
+	page, idx = s.getPageForID(id)
+	if page != nil {
+		err := update(page, idx)
+		s.shardedLocks.Unlock(id)
+		return err
+	}
+	s.shardedLocks.Unlock(id)
+
+	// try again, this time with all locks held
+	s.shardedLocks.LockAll()
+	page, idx = s.getPageForID(id)
+	if page != nil {
+		err := update(page, idx)
+		s.shardedLocks.UnlockAll()
+		return err
+	}
+
+	// page doesn't exist yet, create it
+	pageIdx := id / PageSize
+	page = make([][]T, PageSize)
+	s.cache[pageIdx] = page
+	idx = int(id % PageSize)
+	err := update(page, idx)
+
+	// opportunistically create the next page
+	pageIdx++
+	if len(s.cache) > int(pageIdx) && s.cache[pageIdx] == nil {
+		page = make([][]T, PageSize)
+		s.cache[pageIdx] = page
+	}
+
+	s.shardedLocks.UnlockAll()
+	return err
+}
+
+func (s *paginatedCache[T]) Get(ctx context.Context, id uint64) ([]T, error) {
+	var vec []T
+
+	s.shardedLocks.RLock(id)
+	page, idx := s.getPageForID(id)
+	if page != nil {
+		vec = page[idx]
+	}
+	s.shardedLocks.RUnlock(id)
+
+	if vec != nil {
+		return vec, nil
+	}
+
+	return s.handleCacheMiss(ctx, id)
+}
+
+func (s *paginatedCache[T]) Delete(ctx context.Context, id uint64) {
+	s.shardedLocks.Lock(id)
+	defer s.shardedLocks.Unlock(id)
+
+	if int(id) >= len(s.cache)*PageSize {
+		return
+	}
+
+	page, idx := s.getPageForID(id)
+	if page == nil || page[idx] == nil {
+		return
+	}
+
+	page[idx] = nil
+	atomic.AddInt64(&s.count, -1)
+}
+
+func (s *paginatedCache[T]) handleCacheMiss(ctx context.Context, id uint64) ([]T, error) {
+	vec, err := s.vectorForID(ctx, id)
+	if err != nil {
+		return nil, err
+	}
+
+	atomic.AddInt64(&s.count, 1)
```

Comment:
`handleCacheMiss` is called in `Get` outside of read lock so it could be executed multiple times in parallel for the same id, resulting in multiple `count`'s incrementation for unique vector instead single increment.

Increment can also be moved after `upsert` call to make sure operation succeeded